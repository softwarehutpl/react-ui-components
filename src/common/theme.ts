const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
