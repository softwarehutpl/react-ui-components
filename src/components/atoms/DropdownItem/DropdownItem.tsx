import React, { useContext } from 'react';
import styled from 'styled-components';
import { mapTextPositionToFlexJustify } from '../../../helpers/styleHelperMethods';
import DropdownContext from '../../../context/dropdownContext';
import { COLOR_ALTO } from '../../../common/constants/colors';

interface IDropdownItemProps {
  children?: React.ReactNode;
  className?: string;
  divider?: boolean;
  onClick?: () => void;
}

interface IDropdownItemStyleProps {
  fontColor?: string;
  backgroundColor?: string;
  height?: number;
  padding?: number;
  textAlignment?: 'left' | 'center' | 'right';
  heading?: boolean;
  dividerColor?: string;
  color?: string;
  disabled?: boolean;
}

const DropdownItem = ({ children, className, divider, onClick }: IDropdownItemProps) => {
  if (divider) {
    return <div className={className} />;
  }
  return (
    <div className={className} onClick={onClick}>
      {children}
    </div>
  );
};

const defaultProps = {
  children: null,
  className: '',
  onClick: () => {},
  height: 35,
  padding: 15,
  textAlignment: 'left',
  heading: false,
  divider: false,
  dividerColor: COLOR_ALTO,
  color: 'error',
  disabled: false,
};

DropdownItem.defaultProps = defaultProps;

const StyledDropdownItem = styled(DropdownItem)<IDropdownItemStyleProps>`
  ${({
    theme,
    color = defaultProps.color,
    fontColor,
    backgroundColor,
    height = defaultProps.height,
    padding = defaultProps.padding,
    textAlignment = defaultProps.textAlignment,
    heading = defaultProps.heading,
    divider = defaultProps.divider,
    dividerColor = defaultProps.dividerColor,
    disabled = defaultProps.disabled,
  }) => {
    if (divider) {
      return {
        height: '1px',
        'background-color': dividerColor,
      };
    }
    return {
      color: fontColor || theme.colors[color].base,
      'background-color': backgroundColor || theme.colors[color].light,
      width: '100%',
      height: `${height}px`,
      padding: `0 ${padding}px`,
      'box-sizing': 'border-box',
      display: 'flex',
      'align-items': 'center',
      'justify-content': mapTextPositionToFlexJustify(textAlignment),
      'font-size': heading ? '0.75em' : '1em',
      'font-weight': heading ? 'bold' : 'bolder',
      'text-transform': heading ? 'uppercase' : 'capitalize',
      '&:hover': {
        cursor: 'pointer',
        'background-color': theme.colors[color].dark,
      },
      'pointer-events': disabled ? 'none' : 'auto',
    };
  }}
`;

type DropdownItemProps = IDropdownItemStyleProps & IDropdownItemProps;

const withDropdownContext = (WrappedComponent: React.ComponentType<DropdownItemProps>) => (
  props: DropdownItemProps
) => {
  const { dropdownColor } = useContext(DropdownContext);
  const { color } = props;
  // if color is passed to item, render item with given color, if not, take color from dropdown
  return <WrappedComponent {...props} color={color || dropdownColor} />;
};

export default withDropdownContext(StyledDropdownItem);
