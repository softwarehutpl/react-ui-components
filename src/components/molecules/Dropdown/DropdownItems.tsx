import styled from 'styled-components';
import { defaultProps } from './Dropdown';

export interface IDropdownOptions {
  openingDirection?: 'down' | 'up';
  toggleHeight: number;
  itemsBackgroundColor?: string;
  color?: string;
}

export const DropdownOptions = styled.div<IDropdownOptions>`
  ${({
    openingDirection = defaultProps.openingDirection,
    toggleHeight,
    itemsBackgroundColor,
    theme,
    color = defaultProps.color,
  }) => ({
    position: `absolute`,
    top: openingDirection === 'down' ? `${toggleHeight + 5}px` : 'auto',
    bottom: openingDirection === 'up' ? `${toggleHeight + 5}px` : 'auto',
    left: 0,
    'border-radius': '5px',
    padding: '5px 0',
    'box-shadow': '0 4px 10px 4px rgba(217,217,217,0.6)',
    'background-color': itemsBackgroundColor || theme.colors[color].light,
    width: '100%',
  })}
`;
