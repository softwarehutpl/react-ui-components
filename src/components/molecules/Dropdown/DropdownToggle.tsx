import styled from 'styled-components';
import { defaultProps } from './Dropdown';

export interface IDropdownToggle {
  color?: string;
  width?: number;
  disabled?: boolean;
  backgroundColor?: string;
  padding?: number;
  fontColor?: string;
}

export const DropdownToggle = styled.div<IDropdownToggle>`
  ${({
    theme,
    color = defaultProps.color,
    width = defaultProps.width,
    backgroundColor,
    padding = defaultProps.padding,
    fontColor,
    disabled = defaultProps.disabled,
  }) => {
    let dropdownColor = color;
    if (disabled) {
      dropdownColor = 'disabled';
    }
    return {
      width: `${width}px`,
      'background-color': backgroundColor || theme.colors[dropdownColor].base,
      padding: `${padding}px`,
      color: fontColor || theme.colors[dropdownColor].light,
      'font-weight': 'bold',
      'box-sizing': 'border-box',
      '&:hover': {
        cursor: disabled ? 'default' : 'pointer',
      },
      'text-align': 'left',
    };
  }}
`;
