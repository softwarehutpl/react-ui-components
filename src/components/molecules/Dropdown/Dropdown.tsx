import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { DownArrow } from '@styled-icons/boxicons-solid/DownArrow';
import { UpArrow } from '@styled-icons/boxicons-solid/UpArrow';
import DropdownContext from '../../../context/dropdownContext';
import { DropdownToggle, IDropdownToggle } from './DropdownToggle';
import { DropdownOptions } from './DropdownItems';
import useOutsideClick from '../../../common/hooks/clickOutside';

interface IDropdown extends IDropdownToggle {
  children?: React.ReactNode;
  className?: string;
  title: string;
  itemsBackgroundColor?: string;
  itemsFontColor?: string;
  caretSize?: number;
  openingDirection?: 'down' | 'up';
}

interface IStyledDropdown {
  margin?: number;
}

const arrowStyle = (size?: number) => ({
  width: `${size}px`,
  marginLeft: '10px',
});

const Dropdown = ({
  children,
  color,
  className,
  title,
  width,
  itemsBackgroundColor,
  itemsFontColor,
  openingDirection,
  caretSize,
  disabled,
  backgroundColor,
  padding,
  fontColor,
}: IDropdown) => {
  const [isOpen, setIsOpen] = useState(false);
  const [toggleHeight, setToggleHeight] = useState(0);
  const toggleRef = React.useRef<HTMLDivElement>(null);
  const dropdownRef = React.useRef<HTMLDivElement>(null);

  const toggleDropdown = () => {
    if (!disabled) {
      setIsOpen(!isOpen);
    }
  };

  const closeDropdown = () => {
    setIsOpen(false);
  }

  useOutsideClick(dropdownRef, closeDropdown);

  useEffect(() => {
    if (toggleRef && toggleRef.current) {
      const { height } = toggleRef.current.getBoundingClientRect();
      setToggleHeight(height);
    }
  }, []);

  return (
    <DropdownContext.Provider
      value={{
        dropdownColor: color,
        dropdownWidth: width,
        itemsBackgroundColor,
        itemsFontColor,
        onDropdownItemClick: closeDropdown,
      }}
    >
      <div className={className} ref={dropdownRef}>
        <DropdownToggle
          ref={toggleRef}
          onClick={toggleDropdown}
          color={color}
          width={width}
          backgroundColor={backgroundColor}
          padding={padding}
          fontColor={fontColor}
          disabled={disabled}
        >
          {title}
          {openingDirection === 'up' ? (
            <UpArrow style={arrowStyle(caretSize)} />
          ) : (
            <DownArrow style={arrowStyle(caretSize)} />
          )}
        </DropdownToggle>
        {isOpen && (
          <DropdownOptions
            openingDirection={openingDirection}
            toggleHeight={toggleHeight}
            itemsBackgroundColor={itemsBackgroundColor}
            color={color}
          >
            {children}
          </DropdownOptions>
        )}
      </div>
    </DropdownContext.Provider>
  );
};

export const defaultProps = {
  color: 'primary',
  width: 200,
  padding: 15,
  margin: 0,
  openingDirection: 'down',
  caretSize: 15,
  disabled: false,
};

Dropdown.defaultProps = defaultProps;

const StyledDropdown = styled(Dropdown)<IStyledDropdown>`
  ${({ margin = defaultProps.margin }) => ({
    margin: `${margin}px`,
    position: 'relative',
  })}
`;

export default StyledDropdown;
