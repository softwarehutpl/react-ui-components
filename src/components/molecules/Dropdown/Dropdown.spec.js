import React from 'react';
import { create, act } from 'react-test-renderer';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import Dropdown from './Dropdown';
import DropdownItem from '../../atoms/DropdownItem/DropdownItem';
import theme from '../../../common/theme';
import {DropdownToggle} from './DropdownToggle';

describe('Dropdown component', () => {
  it('should render correctly', () => {
    const tree = create(
      <ThemeProvider theme={theme}>
        <Dropdown title="title">
          <DropdownItem>item</DropdownItem>
        </Dropdown>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot()
  })
  it('should render dropdown with passed title', () => {
    const tree = create(
      <ThemeProvider theme={theme}>
        <Dropdown color="secondary" title="title">
          <DropdownItem>item</DropdownItem>
        </Dropdown>
      </ThemeProvider>
    );
    const instance = tree.root;
    const dropdownToggle = instance.findByType(DropdownToggle)
    expect(dropdownToggle.props.children[0]).toBe('title')
  });

  it('should open dropdown after toggle click', () => {
    const tree = create(
      <ThemeProvider theme={theme}>
        <Dropdown title="title">
          <DropdownItem>item</DropdownItem>
        </Dropdown>
      </ThemeProvider>
    );
    expect(tree.toJSON().children.length).toEqual(1);
    const instance = tree.root;
    const toggle = instance.findByType(DropdownToggle);
    act(() => {
      toggle.props.onClick();
    })
    expect(tree.toJSON().children.length).toEqual(2);
  })

  it('should not open disabled dropdown after toggle click', () => {
    const tree = create(
      <ThemeProvider theme={theme}>
        <Dropdown title="title" disabled>
          <DropdownItem>item</DropdownItem>
        </Dropdown>
      </ThemeProvider>
    );
    expect(tree.toJSON().children.length).toEqual(1);
    const instance = tree.root;
    const toggle = instance.findByType(DropdownToggle);
    act(() => {
      toggle.props.onClick();
    })
    expect(tree.toJSON().children.length).toEqual(1);
  })

  it('should render dropdown with default style', () => {
    const tree = create(
      <ThemeProvider theme={theme}>
        <Dropdown title="title">
          <DropdownItem>item</DropdownItem>
        </Dropdown>
      </ThemeProvider>
    );
    const dropdown = tree.toJSON().children[0];
    expect(dropdown).toHaveStyleRule('width', '200px');
    expect(dropdown).toHaveStyleRule('background-color', '#000000');
    expect(dropdown).toHaveStyleRule('padding', '15px');
    expect(dropdown).toHaveStyleRule('color', '#ffffff');
  });

  it('should render dropdown with custom style', () => {
    const tree = create(
      <ThemeProvider theme={theme}>
        <Dropdown title="title" width={300} backgroundColor="red" padding={10} fontColor="red" margin={10}>
          <DropdownItem>item</DropdownItem>
        </Dropdown>
      </ThemeProvider>
    );
    const dropdown = tree.toJSON().children[0];
    expect(dropdown).toHaveStyleRule('width', '300px');
    expect(dropdown).toHaveStyleRule('background-color', 'red');
    expect(dropdown).toHaveStyleRule('padding', '10px');
    expect(dropdown).toHaveStyleRule('color', 'red');
  });
});
