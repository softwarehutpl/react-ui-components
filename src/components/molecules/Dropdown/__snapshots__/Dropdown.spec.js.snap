// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component should render correctly 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 {
  width: 200px;
  background-color: #000000;
  padding: 15px;
  color: #ffffff;
  font-weight: bold;
  box-sizing: border-box;
  text-align: left;
}

.c1:hover {
  cursor: pointer;
}

.c0 {
  margin: 0px;
  position: relative;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="primary"
    disabled={false}
    onClick={[Function]}
    width={200}
  >
    title
    <svg
      aria-hidden="true"
      className="c2"
      fill="currentColor"
      focusable="false"
      style={
        Object {
          "marginLeft": "10px",
          "width": "15px",
        }
      }
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11.178 19.569a.998.998 0 001.644 0l9-13A.999.999 0 0021 5H3a1.002 1.002 0 00-.822 1.569l9 13z"
      />
    </svg>
  </div>
</div>
`;
